File: backend/app/schemas/calendar.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


class CampusResponse(BaseModel):
    id: int
    code: str
    name: str

    class Config:
        from_attributes = True


class TimelineSlot(BaseModel):
    start: str
    end: Optional[str] = None
    title: str


class BghCalendarPlan(BaseModel):
    event_id: Optional[int] = None
    document_id: int
    plan_name: str
    date: Optional[str] = None
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    location: Optional[str] = None
    timeline: Optional[List[TimelineSlot]] = None
    campuses: List[str]
    is_continuation: bool = False
    event_end_date: Optional[str] = None
    needs_review: bool = False
    source: str = "ai"


class BghCalendarResponse(BaseModel):
    scheduled_plans: List[BghCalendarPlan]
    unscheduled_plans: List[BghCalendarPlan]
    day_counts: dict[str, int]


class PlanEventUpdateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    starts_at: datetime
    ends_at: Optional[datetime] = None
    location: Optional[str] = Field(None, max_length=300)
    timeline: Optional[List[TimelineSlot]] = None
    include_in_calendar: bool = True


class PlanEventCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    starts_at: datetime
    ends_at: Optional[datetime] = None
    location: Optional[str] = Field(None, max_length=300)
    timeline: Optional[List[TimelineSlot]] = None


class PlanEventResponse(BaseModel):
    id: Optional[int] = None
    document_id: int
    title: Optional[str] = None
    location: Optional[str] = None
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None
    timeline: Optional[List[TimelineSlot]] = None
    source: Optional[str] = None
    needs_review: bool = False
    message: str = "Đã cập nhật sự kiện"
    removed: bool = False

File: frontend/src/api/calendar.ts
import api from './client';

export interface Campus {
  id: number;
  code: string;
  name: string;
}

export interface BghCalendarPlan {
  event_id?: number | null;
  document_id: number;
  plan_name: string;
  date: string | null;
  start_time: string | null;
  end_time?: string | null;
  campuses: string[];
  is_continuation?: boolean;
  event_end_date?: string | null;
  needs_review?: boolean;
  source?: string;
}

export interface BghCalendarData {
  scheduled_plans: BghCalendarPlan[];
  unscheduled_plans: BghCalendarPlan[];
  day_counts: Record<string, number>;
}

export interface PlanEventPayload {
  title: string;
  starts_at: string;
  ends_at?: string | null;
}

export interface PlanEventResult {
  id: number;
  document_id: number;
  title: string;
  starts_at?: string | null;
  ends_at?: string | null;
  source: string;
  needs_review: boolean;
  message: string;
}

export const calendarApi = {
  getBghCalendar: async (params: {
    start_date: string;
    end_date: string;
    campus_id?: number;
  }): Promise<BghCalendarData> => {
    const res = await api.get('/tasks/bgh-calendar', { params });
    return res.data;
  },
  updatePlanEvent: async (eventId: number, data: PlanEventPayload): Promise<PlanEventResult> => {
    const res = await api.patch(`/documents/plan-events/${eventId}`, data);
    return res.data;
  },
  createPlanEvent: async (documentId: number, data: PlanEventPayload): Promise<PlanEventResult> => {
    const res = await api.post(`/documents/${documentId}/plan-events`, data);
    return res.data;
  },
};
